docs/enhancements/install-strategy.md

```diff
@@ -0,0 +1,196 @@
+
+These interface conditions will be copied back onto ClusterDeployment by Hive controllers with a "Provision" prefix, the install strategy controller does not need to worry about this and in general should never write to ClusterDeployment. example: Failed becomes ProvisionFailed (which exists today on ClusterDeployment for this purpose)
+
+Install strategy controllers can add their own conditions, but these will not transfer back to the ClusterDeployment. UIs around this process may need to show install strategy conditions explicitly.
+
+The core install strategy CRDs and controllers discussed in this document should live in Hive. However, in theory, it would be possible for an external application to implement it's own install strategy outside of Hive and still interface with ClusterDeployment using this system.
+
+### Migration Plan
+
+No breakage for current installation API is planned. (with exception of as yet unreleased Agent install strategy)
+
+Initial goal would be to breakout ClusterDeployment.Spec.InstallStrategy.Agent. The only implementation here is agent based and this is hidden behind an alpha feature gate, and not yet released to customers. Replace this with a new hive.openshift.io.AgentInstallStrategy CRD, and a ClusterDeployment.Spec.InstallStrategyRef link. Refactor the Agent controllers (currently living in the assisted service repo, coming to hive in future) to watch this resource instead of ClusterDeployment.
+
+In the future, we could migrate the current standalone openshift-install implementation to an OpenShiftInstallStrategy, which continues to use ClusterProvision for it's retry mechanism. To do this we would allow a parallel path to specify where you cannot use both installStrategyRef and provisioning sections together. Provisioning then becomes deprecated and after some period of time, could be removed.
+
+We could add an additional strategy which offers InstallConfig generation, and could be used by users who are ok with a simpler interface and don't want to specify an install config secret.
+
+HypershiftInstallStrategy would be another likely future implementation.
+
+### Examples
+
+```yaml
+apiVersion: hive.openshift.io/v1
+kind: AgentInstallStrategy
+metadata:
+  name: mycluster
+  namespace: hive
+spec:
+  stage: RetryProvisioning
+  clusterDeploymentRef:
+    name: mycluster
+  imageSetRef:
+    openshift-4.6
+  clusterMetadata:
+    # set as soon as known even if install is not completed:
+    adminKubeconfigSecretRef:
+      name: mycluster-3-2sw49-admin-kubeconfig
+    adminPasswordSecretRef:
+      name: mycluster-3-2sw49-admin-password
+    clusterID: 48dec7cf-0c4b-44c6-b840-5c8a6ec93aa6
+    infraID: mycluster-4n2t5
+
+  # agent specific
+  networking:
+    machineNetwork: ...
+  sshPublicKey: ...
+  provisionRequirements:
+    controlPlaneAgents: 3
+    workerAgents: 5
+  controlPlane:
+    hyperthreading: Disabled
+  compute:
+    hyperthreading: Disabled
```

Comment:
Does putting settings like this here in an agent-specific structure just avoid the fact that openshift needs a k8s-native way to express some items that are currently only in the InstallConfig? There's nothing agent-specific about this setting AFAIK.

Reply:
This is one of the hardest parts of the Hive api, we're in between a rock and a hard place all the time with this concept. We require installconfig for normal provisions. Agent provisions explicitly don't want to require an install config. We'd like users to not have to always supply one, but then we're duplicating install config, and boxing ourselves in from using new installer features without a hive update.

This overall proposal is an attempt to deal with that by now leaving it up to install implementations. Because Agent installs want to treat install config as an internal impl detail, they have the flexibility to duplicate the settings they wish to expose without impacting or confusing other parts of the HIve API. In the future, maybe there's a new Install impl that duplicates some of installconfig but gives you a purely CRD defined way to provision with openshift-install and you no longer have to provide the separate secret.